pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,647 @@
+		t := nextScheduledTimeDuration(sched, now)
+		return cj, t, nil
+	}
+
+	scheduledTime := times[len(times)-1]
+	tooLate := false
+	if cj.Spec.StartingDeadlineSeconds != nil {
+		tooLate = scheduledTime.Add(time.Second * time.Duration(*cj.Spec.StartingDeadlineSeconds)).Before(now)
+	}
+	if tooLate {
+		klog.V(4).InfoS("Missed starting window", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()))
+		jm.recorder.Eventf(cj, corev1.EventTypeWarning, "MissSchedule", "Missed scheduled time to start a job: %s", scheduledTime.Format(time.RFC1123Z))
+
+		// TODO: Since we don't set LastScheduleTime when not scheduling, we are going to keep noticing
+		// the miss every cycle.  In order to avoid sending multiple events, and to avoid processing
+		// the cj again and again, we could set a Status.LastMissedTime when we notice a miss.
+		// Then, when we call getRecentUnmetScheduleTimes, we can take max(creationTimestamp,
+		// Status.LastScheduleTime, Status.LastMissedTime), and then so we won't generate
+		// and event the next time we process it, and also so the user looking at the status
+		// can see easily that there was a missed execution.
+		t := nextScheduledTimeDuration(sched, now)
+		return cj, t, nil
+	}
+	if isJobInActiveList(&batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      getJobName(cj, scheduledTime),
+			Namespace: cj.Namespace,
+		}}, cj.Status.Active) || cj.Status.LastScheduleTime.Equal(&metav1.Time{Time: scheduledTime}) {
+		klog.V(4).InfoS("Not starting job because the scheduled time is already processed", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()), "schedule", scheduledTime)
+		t := nextScheduledTimeDuration(sched, now)
+		return cj, t, nil
+	}
+	if cj.Spec.ConcurrencyPolicy == batchv1beta1.ForbidConcurrent && len(cj.Status.Active) > 0 {
+		// Regardless which source of information we use for the set of active jobs,
+		// there is some risk that we won't see an active job when there is one.
+		// (because we haven't seen the status update to the SJ or the created pod).
+		// So it is theoretically possible to have concurrency with Forbid.
+		// As long the as the invocations are "far enough apart in time", this usually won't happen.
+		//
+		// TODO: for Forbid, we could use the same name for every execution, as a lock.
+		// With replace, we could use a name that is deterministic per execution time.
+		// But that would mean that you could not inspect prior successes or failures of Forbid jobs.
+		klog.V(4).InfoS("Not starting job because prior execution is still running and concurrency policy is Forbid", "cronjob", klog.KRef(cj.GetNamespace(), cj.GetName()))
+		t := nextScheduledTimeDuration(sched, now)
+		return cj, t, nil
+	}
+	if cj.Spec.ConcurrencyPolicy == batchv1beta1.ReplaceConcurrent {
+		for _, j := range cj.Status.Active {
+			klog.V(4).InfoS("Deleting job that was still running at next scheduled start time", "job", klog.KRef(j.Namespace, j.Name))
```

Comment:
definitely event worthy.

Reply:
> definitely event worthy.

maybe the delete helper already events?

Reply:
yes, the deleteJob has an event

Reply:
https://github.com/kubernetes/kubernetes/blob/d91a1f7e92a0016e2394a73fd9602b8ae08a2a57/pkg/controller/cronjob/cronjob_controller.go#L381